### ISSUE TITLE
User.Identity.Name empty when implementing Azure B2C in Blazor WebAssembly

### ISSUE THREAD

**Issue by IIStevowII** [dotnet/AspNetCore.Docs#17598]
This issue is about the following doc page:
[Secure an ASP.NET Core Blazor WebAssembly hosted app with Azure Active Directory B2C](https://docs.microsoft.com/en-us/aspnet/core/security/blazor/webassembly/hosted-with-azure-active-directory-b2c?view=aspnetcore-3.1)

## Description
When trying to access *HttpContext.User.Identity.Name* in the Server project inside an API that is protected by the [Authorize] attribute, the name will appear empty. This while I am logged in at the name is shown in the Client application just fine.

## Fix
I had to add the following snippet in Startup.cs to fix this.

```C#
services.AddAuthentication(AzureADB2CDefaults.JwtBearerAuthenticationScheme)
		.AddAzureADB2CBearer(options =&gt; Configuration.Bind("AzureAd", options));

services.Configure&lt;JwtBearerOptions&gt;(AzureADB2CDefaults.JwtBearerAuthenticationScheme, options =&gt;
{
	options.TokenValidationParameters = new Microsoft.IdentityModel.Tokens.TokenValidationParameters()
	{
		NameClaimType = "name"
	};
});
```

I am not sure if this is the right place to report it, not even sure if it is something that should be fixed. At most it will hopefully help other people that come across the same issue and save them some time.

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 7d682d4c-86b5-c47e-89a3-3684b2235320
* Version Independent ID: a80e115c-466a-288d-a918-9ef7514f5add
* Content: [Secure an ASP.NET Core Blazor WebAssembly hosted app with Azure Active Directory B2C](https://docs.microsoft.com/en-us/aspnet/core/security/blazor/webassembly/hosted-with-azure-active-directory-b2c?view=aspnetcore-3.1#feedback)
* Content Source: [aspnetcore/security/blazor/webassembly/hosted-with-azure-active-directory-b2c.md](https://github.com/dotnet/AspNetCore.Docs/blob/master/aspnetcore/security/blazor/webassembly/hosted-with-azure-active-directory-b2c.md)
* Product: **aspnet-core**
* Technology: **aspnetcore-security**
* GitHub Login: @guardrex
* Microsoft Alias: **riande**

**Comment by IIStevowII** [comment]
Just tried the following, still does not seem to work:

```
services.AddAuthentication(AzureADB2CDefaults.BearerAuthenticationScheme)
		.AddAzureADB2CBearer(options => Configuration.Bind("AzureAd", options));

services.Configure<IdentityOptions>(options =>
{
	options.ClaimsIdentity.UserIdClaimType = ClaimTypes.NameIdentifier;
	options.ClaimsIdentity.UserNameClaimType = "name";
});
```

If I pause the debugger on HttpContext.User.Identity, i see the following:
![afbeelding](https://user-images.githubusercontent.com/4059224/78399097-8e69f480-75f4-11ea-8235-8e078a54112c.png)

If I zoom in on the claims, you can see the name identifier in the claim:
![afbeelding](https://user-images.githubusercontent.com/4059224/78399120-9fb30100-75f4-11ea-8d78-bb003c33e3ea.png)

**Comment by guardrex** [comment]
Thanks @IIStevowII ... I'm planning on looking closer at this shortly, over the next few days. I'll get back to you.

**Comment by IIStevowII** [comment]
Sure, thanks! If you need additional info let me know. I am using nightly builds from the special nuget feed, I don't know if that might be an issue? Just to be sure I'll add my current package configuration below.

### Client project
| Package                                               | Version                |
| ----------------------------------------------------- | ---------------------- |
| Microsoft.AspNetCore.Components.Authorization         | 3.1.3                  |
| Microsoft.AspNetCore.Components.WebAssembly           | 3.2.0-preview4.20181.2 |
| Microsoft.AspNetCore.Components.WebAssembly.Build     | 3.2.0-preview4.20181.2 |
| Microsoft.AspNetCore.Components.WebAssembly.DevServer | 3.2.0-preview4.20201.3 |
| Microsoft.Authentication.WebAssembly.Msal             | 3.2.0-preview4.20201.3 |

### Server project
| Package                                            | Version                |
| -------------------------------------------------- | ---------------------- |
| Microsoft.AspNetCore.Authentication.AzureADB2C.UI  | 3.1.3                  |
| Microsoft.AspNetCore.AzureKeyVault.HostingStartup  | 2.0.4                  |
| Microsoft.AspNetCore.Components.WebAssembly.Server | 3.2.0-preview4.20201.3 |

**Comment by guardrex** [comment]
btw and unrelated to the issue at hand ...

Can you let me know which format of scope URI works for you? Which of the following formats are you using ...

1. With the `api://` scheme ...

   ```csharp
   options.ProviderOptions.DefaultAccessTokenScopes.Add("api://{GUID}/{SCOPE NAME}");
   ```

1. No scheme (`api://`) (as the topic currently indicates) ...

   ```csharp
   options.ProviderOptions.DefaultAccessTokenScopes.Add("{GUID}/{SCOPE NAME}");
   ```

1. As presented in the Azure portal to at least some users (me, for example) ...

   ```csharp
   options.ProviderOptions.DefaultAccessTokenScopes.Add("https://XXXXX.onmicrosoft.com/{GUID}/{SCOPE NAME}");
   ```

... where `{GUID}` is probably (by default) the API app Client ID.

**Comment by IIStevowII** [comment]
Currently using option 3. I did specify a custom name for the `App ID URI` property, so it was a bit confusing that the docs kept talking about a GUID. Soon figured out it was about the below value, which seems to be a GUID if not specified.

![afbeelding](https://user-images.githubusercontent.com/4059224/78408761-40f78280-7608-11ea-8161-9c2fbe1733dd.png)

**Comment by guardrex** [comment]
Thanks ... and that's the exact value that you're using for `options.ProviderOptions.DefaultAccessTokenScopes.Add`? ... the full URI + scope?

WRT the "GUID" (aka the client id by default), it's because the portal defaults to the GUID (at least for me it did), and I assume that for a demo that it's ok to just accept whatever the portal provides.

The GUID language might change in the next round of language updates. The WASM security node topics are under heavy revision for improvements before we get to the 3.2 release in May.

Right now, I'm more concerned about what the portal shows relative to what's set in `options.ProviderOptions.DefaultAccessTokenScopes.Add`. When I tried to use the exact full URI from the portal (option 3), it :boom: on me and errored out. When I used option 2, it worked. I'll need to test again from scratch, and I'll probably do that tomorrow morning.

**Comment by IIStevowII** [comment]
Yeah that's the exact value. (It was still hardcoded 🤐)

Fair enough about the GUID being default and therefore used in the demo. Could have sworn it was empty by default, but that was probably an oversight by me. So just to help you out I created a new application just to test things.

### In the portal I used App Registrations (Preview)
1) As you can see in the screenshot, after registering the application I see "Add xxxx" in the overview. That's where I set the Application ID URI (and therefor didn't see a GUID anywhere)
![afbeelding](https://user-images.githubusercontent.com/4059224/78412711-a5204380-7614-11ea-878e-758e617cf3e9.png)

2) If I don't do that this time and proceed to go to "Expose an API" just like the doc says, I get the following popup on the right of my screen **before** I can setup an API (there is a GUID pre-filled this time)
![afbeelding](https://user-images.githubusercontent.com/4059224/78412808-0b0ccb00-7615-11ea-9b3d-f71f237302e1.png)

3) If I then proceed to add API access I see the entire URL below the input field
![afbeelding](https://user-images.githubusercontent.com/4059224/78412862-46a79500-7615-11ea-8fcb-2202a09f7c17.png)

### But..... If I use *Applications -> Add* I get the following screen:
![afbeelding](https://user-images.githubusercontent.com/4059224/78412951-a69e3b80-7615-11ea-85b2-8c5bac466edb.png)

As you can see the *App ID Uri* is not pre-filled with a GUID.
My thought (and probably behavior at that time) would be to put something URL and read friendly there, like the application name for example (or github-test in this example). Can be confusing to associate a part of an URL with a GUID later on, but that could also just be me.

**Comment by guardrex** [comment]
Interesting ... and a little harder to document cleanly since the behavior of it prefilling a client id or not varies like that.

I think dropping "GUID" language makes sense. The topic should probably just refer to the "App ID URI" (whatever it is) and leave it at that. Turns out that Javier has a different portal that allows him to change the scheme. He can make `{ANY SCHEME}://{ANYTHING}/{SCOPE NAME}`. That could be due to an internal portal for MS cats, a premium portal, a preview portal, or an 👽 portal! ... *They are among us!* :smile:

As for `options.ProviderOptions.DefaultAccessTokenScopes.Add`, that's my 😵  **_pain point_** 😵 right now. I'll try again in the morning. Hopefully this time, it will accept the full App ID URI + scope name and just work. If so, then I'll do a quick update on the doc tomorrow morning. If not tho, I'm stuck ... I'll need to work with an Azure engineer on #17304 to figure out what's what.

Thanks for your help. I'll report back on this issue (the one you opened) after research and further testing. If I don't look tomorrow, I'll probably get back to this early next week.

Thanks again @IIStevowII! 🎷

**Comment by guardrex** [comment]
So, I misspoke/misremembered that AAD B2C wasn't exhibiting the strange scope URI behavior. It was just AAD. When using the full API ID URI, the request comes back 401 with ...

> Bearer error="invalid_token", error_description="The audience 'https://XXXXX.onmicrosoft.com/{GUID}' is invalid"

It wasn't a total waste of time discussing it for B2C tho: You confirmed the topic is correct, and you pointed out that depending on where the App ID URI is set it may or may not have a default Server App Client Id sitting there by default.

I also checked the language in both topics, and one strange instruction must remain ... `--app-id-uri "{SERVER API APP CLIENT ID}"` ... that's really confusing. The construction of the full scope URI for the app (either Hosted AAD or Hosted AAD B2C) is created correctly from that setting, but the Client Id is clearly **_not_** the App ID URI as shown in the portal. Even running the command produces an entry of ...

```csharp
options.ProviderOptions.DefaultAccessTokenScopes.Add("api://{SERVER API APP CLIENT ID}/{DEFAULT SCOPE}");
```

... and `api://` must be stripped off to make it work or else I get ...

> The resource principal named api://e5f32fcf-e57c-409a-a2f3-daaf2beec11d was not found

The only way to make it work that I've found is `{API APP CLIENT ID}/{SCOPE NAME}` (when using the Client Id in the App ID URI). I'll have to take that up with Javier and an Azure engineer on the other issue.

**To the matter here ...**

1. Hosted with AAD

   By default, the `User.Identity.Name` is populated, but it receives the value from the `http://schemas.xmlsoap.org/ws/2005/05/identity/claims/name` claim type (e.g., `{GUID}@XXXXX.onmicrosoft.com`).

   Your approach works and pipes the actual name through (e.g., `Chris Green`) to `User.Identity.Name`. The only delta is that one would use `AzureADDefaults.JwtBearerAuthenticationScheme` to get the scheme to bind the options against.

   Default Identity isn't configured ... `UserManager` doesn't work today using the doc's guidance.

1. Hosted with AAD B2C

   By default, unpopulated as you say. Your approach works and pipes the name through (e.g., `Chris Green`).

   Default Identity isn't configured ... `UserManager` doesn't work today using the doc's guidance.

1. Hosted with Identity Server

   There is a default `IdentityUser` (as `ApplicationUser`), so the server app can use `UserManager`. One can get the `IdentityUser` properties (e.g., username, phone, email, etc.) by injecting `UserManager`. Identity Server has a *Profile Service*, but I haven't tried to use it to obtain other profile info. I see a log entry saying that the Profile Service has a `name` claim in it, but I haven't investigated further if using your approach somehow linked to that claim would work to populate `User.Identity.Name`.

Ok ... so I'm going to disable the `UserManager` coverage in the AAD and AAD B2C cases for now until I work out the Server API Identity config. I'll add your solution for obtaining the user's name to the AAD and AAD B2C topics. I'll ping u on the PR.
